src/mo_types/type.ml

```diff
@@ -1059,8 +1074,17 @@ and rel_fields d rel eq tfs1 tfs2 =
   | tf1::tfs1', tf2::tfs2' ->
     (match compare_field tf1 tf2 with
     | 0 ->
-      rel_typ d rel eq tf1.typ tf2.typ &&
-      rel_fields d rel eq tfs1' tfs2'
+      let is_rel =
+        rel_typ d rel eq tf1.typ tf2.typ &&
+        rel_fields d rel eq tfs1' tfs2'
+      in
+      if !Mo_config.Flags.typechecker_combine_srcs && is_rel then begin
+        let srcs = Region_set.union tf1.src.srcs tf2.src.srcs in
+        if rel == eq then
+          tf1.src.srcs <- srcs;
+        tf2.src.srcs <- srcs;
```

Comment:
This mutation scares me. What if the subtyping relation fails to hold after all, won't the destructive merges remain?

Reply:
Maybe a safer approach would be to accumulate the updates in delayed form and then only commit them when the outer call to subtyping succeeds and discard otherwise. Something like a mutable list of thunks that get forced only on success.

Reply:
Good catch, the subtyping relation may fail at a later point, which I didn't think about. I've applied your idea. ([Link to changes](https://github.com/serokell/motoko/pull/57))